File: src/gapscope/gaps.py
from datetime import datetime

from .denoise import apply_denoise
from .models import Baseline, Evidence, GapItem

# Ranking weights: frequency (how many engineers use it) is the primary signal;
# recency boosts but must not zero out a popular-but-older tool. Additive, not
# multiplicative, precisely to avoid that zeroing.
FREQ_WEIGHT = 0.7
RECENCY_WEIGHT = 0.3


def compute_gaps(target_repos, baseline: Baseline, denoise_rules: dict | None = None) -> list[GapItem]:
    """Gap = tools in targets' repos that are not in the baseline.
    Ranked by FREQ_WEIGHT*normalize(frequency) + RECENCY_WEIGHT*normalize(recency)."""
    base = set(baseline.tools)

    agg: dict[str, dict] = {}
    for repo in target_repos:
        tools = apply_denoise(repo.tools, denoise_rules) if denoise_rules else repo.tools
        for tool in tools:
            tool = tool.lower()
            if tool in base:
                continue
            entry = agg.setdefault(tool, {"owners": set(), "repos": [], "recent": repo.pushed_at})
            entry["owners"].add(repo.owner)
            entry["repos"].append(repo.full_name)
            if repo.pushed_at > entry["recent"]:
                entry["recent"] = repo.pushed_at

    if not agg:
        return []

    max_freq = max(len(e["owners"]) for e in agg.values())
    recents: list[datetime] = [e["recent"] for e in agg.values()]
    min_t, max_t = min(recents), max(recents)
    span = (max_t - min_t).total_seconds()

    items: list[GapItem] = []
    for tool, entry in agg.items():
        freq = len(entry["owners"])
        norm_freq = freq / max_freq
        norm_recency = 1.0 if span == 0 else (entry["recent"] - min_t).total_seconds() / span
        rank = round(FREQ_WEIGHT * norm_freq + RECENCY_WEIGHT * norm_recency, 4)
        evidence = [Evidence(repo=r, signal="dependency") for r in sorted(set(entry["repos"]))[:5]]
        items.append(
            GapItem(
                id=tool,
                name=tool,
                kind="tool",
                frequency=freq,
                recencyScore=round(norm_recency, 4),
                rankScore=rank,
                evidence=evidence,
            )
        )

    items.sort(key=lambda g: (g.rankScore, g.frequency), reverse=True)
    return items

File: web/lib/engine/gaps.ts
import type { Gap } from "../types";
import type { ExtractedRepo } from "./types";
import { applyDenoise, type DenoiseRules } from "./denoise";

export function round4(x: number): number {
  return Math.round(x * 1e4) / 1e4;
}

const FREQ_WEIGHT = 0.7;
const RECENCY_WEIGHT = 0.3;

type Accumulator = {
  owners: Set<string>;
  repos: string[];
  recent: Date;
};

export function computeGaps(
  targetRepos: ExtractedRepo[],
  baselineTools: string[],
  denoiseRules?: DenoiseRules,
): Gap[] {
  const base = new Set(baselineTools.map((t) => t.toLowerCase()));

  const agg = new Map<string, Accumulator>();

  for (const repo of targetRepos) {
    const tools = denoiseRules ? applyDenoise(repo.tools, denoiseRules) : repo.tools;
    for (const rawTool of tools) {
      const tool = rawTool.toLowerCase();
      if (base.has(tool)) continue;
      const existing = agg.get(tool);
      if (!existing) {
        agg.set(tool, { owners: new Set([repo.owner]), repos: [repo.fullName], recent: repo.pushedAt });
      } else {
        existing.owners.add(repo.owner);
        existing.repos.push(repo.fullName);
        if (repo.pushedAt > existing.recent) {
          existing.recent = repo.pushedAt;
        }
      }
    }
  }

  if (agg.size === 0) return [];

  const maxFreq = Math.max(...Array.from(agg.values()).map((e) => e.owners.size));

  const recents = Array.from(agg.values()).map((e) => e.recent.getTime());
  const minT = Math.min(...recents);
  const maxT = Math.max(...recents);
  const span = maxT - minT;

  const items: Gap[] = [];

  for (const [tool, entry] of agg) {
    const freq = entry.owners.size;
    const normFreq = freq / maxFreq;
    const normRecency = span === 0 ? 1.0 : (entry.recent.getTime() - minT) / span;
    const rankScore = round4(FREQ_WEIGHT * normFreq + RECENCY_WEIGHT * normRecency);
    const recencyScore = round4(normRecency);

    const uniqueRepos = Array.from(new Set(entry.repos)).sort();
    const evidence = uniqueRepos.slice(0, 5).map((repo) => ({ repo, signal: "dependency" as const }));

    items.push({
      id: tool,
      name: tool,
      kind: "tool",
      frequency: freq,
      recencyScore,
      rankScore,
      evidence,
      docs: [],
      projects: { small: [], big: [] },
      research: {},
    });
  }

  items.sort((a, b) => {
    if (b.rankScore !== a.rankScore) return b.rankScore - a.rankScore;
    return b.frequency - a.frequency;
  });

  return items;
}
